daemon/graphdriver/copy/copy.go

```diff
@@ -106,15 +114,131 @@ func copyXattr(srcPath, dstPath, attr string) error {
 	return nil
 }
 
-// DirCopy copies or hardlinks the contents of one directory to another,
-// properly handling xattrs, and soft links
-func DirCopy(srcDir, dstDir string, copyMode Mode) error {
+func concurrentCopyHelper(copiedFiles map[fileID]*dstFilePathWithLock, copiedFilesMutex sync.Locker, fileToCopy fileCopyInfo, copyXattrs bool, copyWithFileRange, copyWithFileClone *bool) error {
+	id := fileID{dev: fileToCopy.stat.Dev, ino: fileToCopy.stat.Ino}
+
+	copiedFilesMutex.Lock()
```

Comment:
Can we replace the locking strategy with https://github.com/moby/moby/blob/master/pkg/locker/locker.go

Reply:
We can, but, the sync.Locker would need to change to take in an interface{} because this is a composite structure, or we would have to do a fmt.Sprintf on the above. I'm hesitant to do the latter because this code is already CPU intensive, and we're spending a bunch of time in Go's runtime.

Or we would need to clone one that ould work with FileID.

Reply:
Also, I'm not sure that it provides much value other than abstracting out the copiedFilesMutex, but if you want, I can write a better generic synchronizer.

Reply:
I'm thinking something that's like
```
type Serializer interface {
   call(serializerKey interface{}, cb func())
}
```

Reply:
I was thinking it could just use the file path.

The main benefit is reducing the potential for messing up the locking.

Reply:
It can't use the file path, it has to use the dev + inode number because that's what the hardlink is based on.

Reply:
I'm reluctant to use locker. I wrote this up with a string: https://github.com/sargun/moby/commit/705d334748fbce8557b9dbda2d693c4e0e22b7c7

And it feels awkward, and there's a performance overhead.

I've also refactored the code so that locking is pretty fool proof. The `dstFilePathWithLock` container has a guaranteed unlock path, and explicit interleaving is removed. (Although it's implicit due to the fact it happens on instantiation of a new object).